zkevm-circuits/src/evm_circuit/op_execution/dup.rs

```diff
@@ -0,0 +1,462 @@
+
+    fn construct(case_allocations: Vec<CaseAllocation<F>>) -> Self {
+        let [mut success, stack_overflow, stack_underflow, out_of_gas]: [CaseAllocation<F>; 4] =
+            case_allocations.try_into().unwrap();
+        Self {
+            success: DupSuccessAllocation {
+                case_selector: success.selector.clone(),
+                word: success.words.pop().unwrap(),
+                selectors: success.cells.try_into().unwrap(),
+            },
+            stack_overflow: stack_overflow.selector,
+            stack_underflow: stack_underflow.selector,
+            out_of_gas: (
+                out_of_gas.selector,
+                out_of_gas.resumption.unwrap().gas_available,
+            ),
+        }
+    }
+
+    fn constraints(
+        &self,
+        state_curr: &OpExecutionState<F>,
+        state_next: &OpExecutionState<F>,
+    ) -> Vec<Constraint<F>> {
+        let OpExecutionState { opcode, .. } = &state_curr;
+        // use num_duplicated to represents 'x' value of 'dupx'
+        let num_duplicated = opcode.expr() - OpcodeId::DUP1.expr() + 1.expr();
+        // lookup in range 16 for dup
+        let common_lookups = vec![Lookup::FixedLookup(
+            FixedLookup::Range16,
+            [opcode.expr() - OpcodeId::DUP1.expr(), 0.expr(), 0.expr()],
+        )];
+
+        let success = {
+            // interpreter state transition constraints
+            let state_transition_constraints = vec![
+                state_next.global_counter.expr()
+                    - (state_curr.global_counter.expr() + 2.expr()),
+                state_next.program_counter.expr()
+                    - (state_curr.program_counter.expr() + 1.expr()),
+                // dupx contains one time push operation
+                state_next.stack_pointer.expr()
+                    - (state_curr.stack_pointer.expr() - 1.expr()),
+                state_next.gas_counter.expr()
+                    - (state_curr.gas_counter.expr() + GasCost::FASTEST.expr()),
+            ];
+
+            let DupSuccessAllocation {
+                case_selector,
+                word,
+                selectors,
+            } = &self.success;
+
+            let mut dup_constraints = vec![];
+            for idx in 0..16 {
+                // selector can transit from 1 to 0 only once as [1, 1, 1, ...,
+                // 0, 0, 0]
+                if idx > 0 {
+                    let diff =
+                        selectors[idx - 1].expr() - selectors[idx].expr();
+                    dup_constraints.push(diff.clone() * (1.expr() - diff));
+                }
+                // selectors needs to be 0 or 1
+                dup_constraints.push(
+                    selectors[idx].expr() * (1.expr() - selectors[idx].expr()),
+                );
+            }
+
+            let selectors_sum =
+                selectors.iter().fold(0.expr(), |sum, s| sum + s.expr());
+            dup_constraints.push(selectors_sum - num_duplicated.clone());
```

Comment:
What's the purpose of `selectors` here?

Reply:
Doesn't `num_duplicated` already identify that?

Reply:
looks like somewhat redundant indeed.  thanks for catching ,will try deletion to see the result

Reply:
what do you think if push gadget really needs selectors ?

Reply:
That serve as boolean selectors for independent lookups, which we can't only use `num_duplicated` to achieve.